libcrux-ml-dsa/src/hash_functions.rs

```diff
@@ -76,44 +76,78 @@ pub(crate) mod shake128 {
 
 /// A portable implementation of [`shake128::Xof`] and [`shake256::Xof`].
 pub(crate) mod portable {
-    use libcrux_sha3::portable::{
-        incremental::{self, shake128_absorb_final, shake128_init},
-        shake128, shake256, KeccakState,
-    };
+    use libcrux_sha3::portable::incremental;
 
     use super::{shake128, shake256};
 
     /// Portable SHAKE 128 x4 state.
     ///
     /// We're using a portable implementation so this is actually sequential.
+    #[cfg_attr(hax, hax_lib::opaque_type)]
     pub(crate) struct Shake128X4 {
-        state0: KeccakState,
-        state1: KeccakState,
-        state2: KeccakState,
-        state3: KeccakState,
+        state0: libcrux_sha3::portable::KeccakState,
+        state1: libcrux_sha3::portable::KeccakState,
+        state2: libcrux_sha3::portable::KeccakState,
+        state3: libcrux_sha3::portable::KeccakState,
+    }
+
+    fn init_absorb(input0: &[u8], input1: &[u8], input2: &[u8], input3: &[u8]) -> Shake128X4 {
+        let mut state0 = incremental::shake128_init();
+        incremental::shake128_absorb_final(&mut state0, &input0);
+
+        let mut state1 = incremental::shake128_init();
+        incremental::shake128_absorb_final(&mut state1, &input1);
+
+        let mut state2 = incremental::shake128_init();
+        incremental::shake128_absorb_final(&mut state2, &input2);
+
+        let mut state3 = incremental::shake128_init();
+        incremental::shake128_absorb_final(&mut state3, &input3);
+
+        Shake128X4 {
+            state0,
+            state1,
+            state2,
+            state3,
+        }
+    }
+
+    fn squeeze_first_five_blocks(
```

Comment:
Member functions here are not a problem. However, the member function we used here was exposing a type internal to libcrux-sha3 (`Libcrux_sha3.Portable.t_KeccakState`) which breaks the hash-functions abstraction, so we eliminated it. We could have kept it with some sort of "not(hax)" annotation so that it got dropped from the interface, but since we were already moving the parent function out of the impl (also to aid abstraction) it seemed too much to have three levels of functions.

Reply:
Not sure I understand how the two are connected. We have an opaque struct now and no `not(hax)`, so this should work with member functions too. But let's get this in. We can make it nice again later.